src/dfx/src/commands/ledger/mod.rs

```diff
@@ -147,7 +150,6 @@ pub async fn transfer(
                     .context("Failed to decode transfer response.")?;
                 match result {
                     Ok(block_height) => break block_height,
-                    Err(TransferError::TxDuplicate { duplicate_of }) => break duplicate_of,
```

Comment:
Please leave this in: it makes the call idempotent in the presence of fields that detect duplicates.

Reply:
This does mean that the tests will have to go through some extra work to make sure that there were no duplicate transfers.

Reply:
The call is idempotent either way, this just makes sure the user knows that it is idempotent. Otherwise the user can think that 2 transfers are made when only 1 went through.

Reply:
What is the benefit of swallowing the duplicate-transaction-error?

Reply:
Two benefits, one general, one specific:
1. General: If you're retrying, it's because the result of the first call was a failure.  There's a class of failures where an operation succeeds on the server, but communication fails before the caller finds out that the operation succeeded.

In this case, the advantage of "swallowing" the error is that you can make the call a second time, and check for success just as you did the first time.

Put another way, one can loop until `dfx ledger transfer --icp 1 --memo 1 --created-at-time "$t"` returns 0, and know that the operation succeeded exactly once.  Without this "swallowing", one must loop until the call either succeeds, or until it fails with a duplicate error.

2. Specific: `dfx ledger create-canister` [calls transfer_cmc()](https://github.com/dfinity/sdk/blob/master/src/dfx/src/commands/ledger/create_canister.rs#L83) (which in turn [calls transfer()](https://github.com/dfinity/sdk/blob/master/src/dfx/src/commands/ledger/mod.rs#L181)), and then calls notify_create() with the block height in order to obtain and display the canister id. If `transfer()` returns an error, then as currently written, `dfx ledger create-canister` will not call `notify` and will not display the principal of the newly-created canister.

Reply:
Thanks. 

Commit https://github.com/dfinity/sdk/pull/3018/commits/b276bcf8d0eb2c9560e893ba33de5f7615744365 solves number 2.

About number 1, 

> Put another way, one can loop until dfx ledger transfer --icp 1 --memo 1 --created-at-time "$t" returns 0, and know that the operation succeeded exactly once. Without this "swallowing", one must loop until the call either succeeds, or until it fails with a duplicate error.

It may be easier in some scenarios, but can cause confusion in other scenarios. These are two different outcomes that would be blurred into one. A user can send multiple transactions and all would show success but only 1 would go through on the ledger. 

There is a reason the ledger canister returns Err(TransferError::TransactionDuplicate) on a duplicate-transaction. If the ledger would return Ok(height) even on a duplicate transaction it would cause confusion.

Reply:
Yes, there are tradeoffs.  Will discuss.